File: frontend/src/types/scanner.ts
export type ScanStatus = "ok" | "no_instrument" | "no_candles" | "no_indicators" | "error";

export interface ScannerInstrumentRequest {
  ticker: string;
  engine?: string;
  market?: string;
  board?: string;
}

export interface ScannerRequest {
  instruments: ScannerInstrumentRequest[];
  timeframe: string;
  lookback?: number;
}

export interface ScannerRow {
  ticker: string;
  name: string | null;
  engine: string | null;
  market: string | null;
  board: string | null;
  instrument_id: number | null;
  timeframe: string;
  status: ScanStatus;
  last_close: number | null;
  change_percent: number | null;
  aggregate_signal: string | null;
  total_score: number | null;
  confidence: string | null;
  bullish_count: number | null;
  bearish_count: number | null;
  caution_count: number | null;
  rsi: number | null;
  macd_histogram: number | null;
  atr_percent: number | null;
  nearest_support: number | null;
  nearest_resistance: number | null;
  distance_to_support_percent: number | null;
  distance_to_resistance_percent: number | null;
  summary: string | null;
  error: string | null;
}

export interface ScannerResponse {
  timeframe: string;
  rows: ScannerRow[];
  generated_at: string;
}

File: backend/app/schemas/scanner.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel

ScanStatus = Literal["ok", "no_instrument", "no_candles", "no_indicators", "error"]


class ScannerInstrumentRequest(BaseModel):
    ticker: str
    engine: str = "stock"
    market: str = "shares"
    board: str = "TQBR"


class ScannerRequest(BaseModel):
    instruments: list[ScannerInstrumentRequest]
    timeframe: str = "1d"
    lookback: int = 100


class ScannerRow(BaseModel):
    ticker: str
    name: str | None = None
    engine: str | None = None
    market: str | None = None
    board: str | None = None
    instrument_id: int | None = None
    timeframe: str
    status: ScanStatus
    last_close: float | None = None
    change_percent: float | None = None
    aggregate_signal: str | None = None
    total_score: int | None = None
    confidence: str | None = None
    bullish_count: int | None = None
    bearish_count: int | None = None
    caution_count: int | None = None
    rsi: float | None = None
    macd_histogram: float | None = None
    atr_percent: float | None = None
    nearest_support: float | None = None
    nearest_resistance: float | None = None
    distance_to_support_percent: float | None = None
    distance_to_resistance_percent: float | None = None
    summary: str | None = None
    error: str | None = None
    last_timestamp: datetime | None = None


class ScannerResponse(BaseModel):
    timeframe: str
    rows: list[ScannerRow]
    generated_at: datetime
